File: seeker/snippet/combinationSum.java
//date: 2022-06-27T16:52:33Z
//url: https://api.github.com/gists/f909f0dad76638b9a8d928c6b5662e13
//owner: https://api.github.com/users/archana77archana

class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList();
        int start = 0;
        List<Integer> storeCandidates = new ArrayList();
        
        // We will backtracking our way 
        backTrackSolution(candidates, start, target, storeCandidates, result);
        
        return result;
    }
    
    private void backTrackSolution(int[] candidates, int start, int target, List<Integer> storeCandidates, List<List<Integer>> result) {
        // exit scenario
        if(target < 0)
            return;
        
        // Add rest of the candidates to the result
        if(target == 0)
            result.add(new ArrayList(storeCandidates));
        
        for(int i = start; i < candidates.length; i++) {
            // Classic Backtracking!
            storeCandidates.add(candidates[i]);

            backTrackSolution(candidates, i, target-candidates[i], storeCandidates, result);
            
            storeCandidates.remove(storeCandidates.size() - 1);
        }
    }
}

File: seeker/snippet/combination-sum.py
#date: 2023-06-26T16:58:17Z
#url: https://api.github.com/gists/706375d0b18e63ffa906acd8c13fc4b0
#owner: https://api.github.com/users/markbrutx

def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:

    res = []
    def backtrack(remain, cur, start):

        if remain == 0:
            res.append(list(cur))
            return
        elif remain < 0:
            return
        else:
            for i in range(start, len(candidates)):
                cur.append(candidates[i])
                backtrack(remain - candidates[i], cur, i)

                cur.pop()
    backtrack(target, [], 0)
    return res